src/GovUk.Education.ExploreEducationStatistics.Data.Model/Repository/FootnoteRepository.cs

```diff
@@ -78,25 +76,100 @@ public FootnoteRepository(StatisticsDbContext context)
+                .Select(filter => filter.FilterItemId)
+                .ToList();
+
+            var filterGroupsIdsSanitised = filterIdsResult
+                .Select(filter => filter.FilterGroupId)
+                .Distinct()
+                .ToList();
+
+            var filterIdsSanitised = filterIdsResult
+                .Select(filter => filter.FilterId)
+                .Distinct()
+                .ToList();
+
+            // Sanitise the indicator id's to make sure they belong to the subject
+            var indicatorIdsSanitised = await _context.Indicator
+                .Where(indicator => indicatorIds.Contains(indicator.Id)
+                                    && indicator.IndicatorGroup.SubjectId == subjectId)
+                .Select(indicator => indicator.Id)
                 .ToListAsync();
+
+            // A footnote must be linked to the release id
+            var footnotesQueryable = _context
+                .Footnote
+                .Where(footnote => footnote.Releases.Any(releaseFootnote => releaseFootnote.ReleaseId == releaseId));
+
+            // If the list of filter and indicator id's are empty,
+            // return early with only footnotes that apply directly to this subject
+            if (!filterIdsResult.Any() && !indicatorIdsSanitised.Any())
+            {
+                return await footnotesQueryable.Where(footnote =>
+                        footnote.Subjects.Any(subjectFootnote => subjectFootnote.SubjectId == subjectId)
+
+                        // We don't allow a footnote to apply to directly to a subject and to specific subject data
+                        // of the same subject at the same time, so these additional conditions aren't strictly necessary
+                        && footnote.FilterItems.All(itemFootnote =>
+                            itemFootnote.FilterItem.FilterGroup.Filter.SubjectId != subjectId)
+                        && footnote.FilterGroups.All(filterGroupFootnote =>
+                            filterGroupFootnote.FilterGroup.Filter.SubjectId != subjectId)
+                        && footnote.Filters.All(filterFootnote =>
+                            filterFootnote.Filter.SubjectId != subjectId)
+                        && footnote.Indicators.All(indicatorFootnote =>
+                            indicatorFootnote.Indicator.IndicatorGroup.SubjectId != subjectId))
+                    .ToListAsync();
+            }
+
+            // A footnote which applies directly to any subjects must include this subject id
+            footnotesQueryable = footnotesQueryable.Where(footnote =>
+                !footnote.Subjects.Any() || footnote.Subjects.Any(subjectFootnote =>
+                    subjectFootnote.SubjectId == subjectId));
+
+            // A footnote which applies to any indicators of this subject must include one of the indicator id's
+            footnotesQueryable = footnotesQueryable.Where(footnote =>
+                footnote.Indicators.All(indicatorFootnote =>
```

Comment:
Can't quite get my head round this left-hand part of this "or" clause - we're saying to include footnotes where all of the indicators belong to a different subject?

Reply:
Left-hand part:
`footnote.Indicators.All(indicatorFootnote =>indicatorFootnote.Indicator.IndicatorGroup.SubjectId != subjectId)`

A footnote can apply to several subjects.

If all of the footnote indicator links are for other subjects it shouldn't exclude this footnote.
The footnote might apply to this subject in some other way (i.e. apply directly to the subject or apply to filters/filter groups/filters items of this subject)

Then if that's false, i.e. one of the links is for this subject we need to check it's contained in the list of indicator id's (the right-hand part).

I suppose there's a possibility that the footnote doesn't apply to this subject at all in which case the following could happen:

- Matches the left hand part here.
- Matches all the left hand parts below in this code for filters/filter groups/filters items because all of those links are for other subjects too.

and then it would still be included. I'm sure there's a test case for this but I'll double check this isn't a bug!

Reply:
Update: there was a bug here, so good job you questioned this!

I didn't have any examples of footnotes which were only applicable exclusively to other subjects than the one being queried. It turns out that as soon as I added them they were being returned even though they were not relevant.

I had to largely rewrite the `GetFilteredFootnotes` method again to resolve this. It's very gnarly but it's got a lot more comments now.

I've added an entirely new unit test `GetFilteredFootnotes_FilterBySubject` which has two subjects and sets up a bunch of footnotes that apply to one subject or the other, but not both. It queries for `subject1` and checks that `subject2`'s footnotes are excluded. Prior to this change, all of `subject2`'s footnotes would have been returned due to passing all the clauses but never matching on anything.

Changing `GetFilteredFootnotes` then showed up some mistakes I'd got in `GetFilteredFootnotes_FilterByFilterItems` not expecting some footnotes which we should be expecting, so I've corrected that too. 🙂

Reply:
Ahh brill, happy conclusion then :D  Nicely done.